worker/uniter/operation/deploy.go

```diff
@@ -51,6 +51,12 @@ func (d *deploy) Prepare(state State) (*State, error) {
 	if err := d.checkAlreadyDone(state); err != nil {
 		return nil, errors.Trace(err)
 	}
+	if d.kind == Install {
+		// TODO: check for terminal status?
```

Comment:
We should check for terminal status here before removing the upgrade charm profile data. If the status is not terminal, we should return out an error, but we should remove the profile data so that subsequent deploys should proceed.

Reply:
There isn't a good way to determine if a doc in terminal status should exist.  The only time it should exist at this point, is for 2nd unit on a machine, such as a subordinate.  

Maybe checking to see if there is a doc, because the doc should already exist if it's going to.  If the doc exists, check the status for terminal.

Reply:
What happens if we error out here?  Is the uniter restarted?  and Perhaps the change would be complete by that point?

At this point we have the charmURL... could we see if the charm has a profile, and thus a doc to be removed?

Reply:
> What happens if we error out here? Is the uniter restarted?

So, it depends... on [resume install](https://github.com/juju/juju/blob/d865bd728bc8219d410f97b9be4bc49da0167b9a/worker/uniter/uniter.go#L209-L215) it will cause the uniter to return out and not attempt a restart. BUT code later on will cause the uniter to [process the same operation](https://github.com/juju/juju/blob/d865bd728bc8219d410f97b9be4bc49da0167b9a/worker/uniter/uniter.go#L352-L392) and only certain error results will cause termination.

> At this point we have the charmURL... could we see if the charm has a profile, and thus a doc to be removed?

Yes, but I'd probably have a bespoke function in this case.